glommio/src/io/immutable_file.rs

```diff
@@ -0,0 +1,395 @@
+        self.flush_disabled = flush_disabled;
+        self
+    }
+
+    /// Define the buffer size that will be used by the sequential operations on
+    /// this [`ImmutableFile`]
+    ///
+    /// [`ImmutableFile`]: ImmutableFile
+    pub fn with_buffer_size(mut self, buffer_size: usize) -> Self {
+        self.buffer_size = buffer_size;
+        self
+    }
+
+    /// Builds an [`ImmutableFileSink`] with the properties defined by this
+    /// builder.
+    ///
+    /// The resulting sink can be written to until [`seal`] is called.
+    ///
+    /// [`ImmutableFile`]: ImmutableFileSink
+    /// [`new`]: ImmutableFileBuilder::new
+    /// [`seal`]: ImmutableFileSink::seal
+    pub async fn build_sink(self) -> Result<ImmutableFileSink> {
+        println!("siink builder");
+        let file = DmaOpenOptions::new()
+            .read(true)
+            .write(true)
+            .create_new(true)
+            .open(self.path)
+            .await?;
+        let writer = DmaStreamWriterBuilder::new(file)
+            .with_sync_on_close_disabled(self.flush_disabled)
+            .with_buffer_size(self.buffer_size)
+            .with_write_behind(self.concurrency)
+            .build();
+
+        Ok(ImmutableFileSink {
+            writer,
+            max_merged_buffer_size: self.max_merged_buffer_size,
+            max_read_amp: self.max_read_amp,
+        })
+    }
+
+    /// Builds an [`ImmutableFile`] with the properties defined by this
+    /// builder.
+    ///
+    /// The resulting file cannot be written to. Glommio may optimize access
+    /// patterns by assuming the file is read only. Writing to this file
+    /// from an external process leads to undefined behavior.
+    ///
+    /// [`ImmutableFile`]: ImmutableFileSink
+    /// [`new`]: ImmutableFileBuilder::new
+    pub async fn build_existing(self) -> Result<ImmutableFile> {
+        let file = Rc::new(
+            DmaOpenOptions::new()
+                .read(true)
+                .write(false)
+                .open(self.path)
+                .await?,
+        );
+        let stream_builder = DmaStreamReaderBuilder::from_rc(file)
+            .with_buffer_size(self.buffer_size)
+            .with_read_ahead(self.concurrency);
+
+        Ok(ImmutableFile {
+            stream_builder,
+            max_merged_buffer_size: self.max_merged_buffer_size,
+            max_read_amp: self.max_read_amp,
+        })
+    }
+}
+
+impl ImmutableFileSink {
+    /// Seals the file for further writes.
+    ///
+    /// Once this is called, it is no longer possible to write to the resulting
+    /// [`ImmutableFile`]
+    pub async fn seal(mut self) -> Result<ImmutableFile> {
+        let stream_builder = poll_fn(|cx| self.writer.poll_seal(cx)).await?;
+
+        Ok(ImmutableFile {
+            stream_builder,
+            max_merged_buffer_size: self.max_merged_buffer_size,
+            max_read_amp: self.max_read_amp,
+        })
+    }
+}
+
+impl AsyncWrite for ImmutableFileSink {
```

Comment:
Not a fan of the `ImmutableFileSink` construct and this line captures why: implementing `Write` on `Immutable` feels really wrong. Or, at least, the naming is unfortunate.

Could we instead build an immutable file from a regular `DmaFile`/`BufferedFile`? We could have a consuming function like so:

```
pub fn make_immutable(self) -> ImmutableFile {
  // convert and return
}
```

And we could also have a new `open_immutable` variant on the `OpeOptions` struct that would return an `ImmutableFile` directly.

I feel that this would simplify this work a bit as the write portion of your PR largely overlaps with the stream facilities we already have. Correct me if I'm wrong!

Reply:
Not sure what your concern is. We are not implementing `Write` on the ImmutableFile. We are implementing it in the sink.

Suppose you now construct an ImmutableFile from a `DmaFile`. You now have to manually open the file, meaning the flags are whatever you passed it.

That means that if, by an accident, you happen to rewrite a file that was already in existence, the original `ImmutableFile` is changed and the contract of immutability is void.

That doesn't happen with the sink construct, because we force the file to be opened with `O_EXCL`: it will fail if it already exists.

So all in all, the sink is a necessary construct to make sure that the files are immutable as long as a certain path is always accessed through the Immutable Api.

Much harder to mess up.

Conceptually, it also matches the lifetime of phases of a file -> Create -> write, write, write -> seal -> read read read.